Hide empty categories in mod selector overlay

andy840119 opened the pull request:
![image](https://user-images.githubusercontent.com/9100368/59480474-4fb35400-8e93-11e9-9f31-2f174f72d3a2.png)
Before
.
![image](https://user-images.githubusercontent.com/9100368/59480511-71144000-8e93-11e9-8dbd-0f1a5bcebc29.png)
After

andy840119 committed:
hide label when mod is empty

andy840119 committed:
use show/hide instead because FillFlowContainer's spacing

andy840119 committed:
oops

andy840119 committed:
expanded -> expand

andy840119 committed:
remve unnecessary effect.

peppy commented on osu.Game/Overlays/Mods/ModSection.cs:
```diff
@@ -50,6 +50,19 @@ public IEnumerable<Mod> Mods
+                else
+                {
+                    headerLabel.FadeOut(200);
+                    Hide();
```
This hide is instant, meaning you'll never see the `FadeOut`. Maybe just use `this.FadeOut` and remove the `headerLabel` transition in both cases?

This whole addition could be reduced to `this.FadeTo(value.Any() ? 1 : 0, 200);`

peppy replied:
This feedback *still applies*.

You are missing the fact that Hide() is doing an instant hide. The headerLabel cannot fade out as you have specified because it is hidden by the parent element. Please explain your thought process and why you are rejecting the proposal in my above comment.

iiSaLMaN replied:
to be more clear:
`Show()` = `this.FadeIn(0);`
`Hide()` = `this.FadeOut(0);`

andy840119 replied:
Sorry i though `Show()` is not equal to `this.FadeIn(0)` before ,my mistake.

andy840119 replied:
https://streamable.com/almtk
And i found an other issue.
If added a hidden object in `FillFlowContainer` and re-show it, the effect of moving position is not in the right place.

andy840119 commented on osu.Game/Overlays/Mods/ModSelectOverlay.cs:
```diff
@@ -172,6 +172,8 @@ public ModSelectOverlay()
                                     Spacing = new Vector2(0f, 10f),
                                     Width = content_width,
+                                    LayoutDuration = 200,
+                                    LayoutEasing = Easing.OutQuint,
```
@peppy Because of FillFlowContainer's spacing
So i use Show/Hide instead
.
https://streamable.com/m654k

peppy replied:
see my above comment.

andy840119 replied:
oops i already fixed that but forgot to push :P

peppy replied:
Am I waiting on a further commit to re-review this?

andy840119 replied:
You can review it now : )

iiSaLMaN commented:
please change the title to something like `Hide empty categories in mod selector overlay`

andy840119 committed:
using FadeTo() instead of FadeIn()/FadeOut()

andy840119 committed:
using FadeTo instead of show/hide
headerLabel.FadeTo() is still remain because effect can be visible when expand== true

andy840119 committed:
Merge branch 'master' of https://github.com/ppy/osu into andy840119/_20190614_hidden_nod_label_when_empty

iiSaLMaN commented on osu.Game/Overlays/Mods/ModSection.cs:
```diff
@@ -50,6 +50,10 @@ public IEnumerable<Mod> Mods
+
+                var expand = value.Any();
+                headerLabel.FadeTo(expand ? 1 : 0, 200);
+                this.FadeTo(expand ? 1 : 0);
```
https://github.com/ppy/osu/blob/4e85481fd417131c2fd88f288e1ef95624358ae7/osu.Game/Overlays/Mods/ModSection.cs#L54-L56
That can be simplified into `this.FadeTo(value.Any() ? 1 : 0, 200)`, let me explain why, because if you change the alpha value of a container, it will affect its children, and `headerLabel` is a child of this container

andy840119 replied:
But if use `this.FadeTo(value.Any() ? 1 : 0, 200)` instead, `FillFlowContainer`'s `LayoutEasing` will be trigger on 200 ms
.
or maybe use `this.FadeTo(value.Any() ? 1 : 0,value.Any() ? 200 : 0)` ?

iiSaLMaN replied:
we want it to fade in and fade out on 200ms, to give a good look when the mod category shows,  unless there will be something wrong?

andy840119 replied:
https://streamable.com/35f5g
this is the effect when just use `this.FadeTo(value.Any() ? 1 : 0, 200);`

peppy committed:
Merge branch 'master' into andy840119/_20190614_hidden_nod_label_when_empty

peppy committed:
Adjust for readability

peppy committed:
Rename mod select test scene
